cpp/include/raft/core/pinned_mdspan.hpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/memory_type.hpp>
+
+#include <raft/core/host_device_accessor.hpp>
+
+namespace raft {
+
+template <typename AccessorPolicy>
+using pinned_accessor = host_device_accessor<AccessorPolicy, memory_type::pinned>;
+
+/**
+ * @brief std::experimental::mdspan with pinned tag to avoid accessing incorrect memory location.
+ */
+template <typename ElementType,
+          typename Extents,
+          typename LayoutPolicy   = layout_c_contiguous,
+          typename AccessorPolicy = std::experimental::default_accessor<ElementType>>
+using pinned_mdspan = mdspan<ElementType, Extents, LayoutPolicy, pinned_accessor<AccessorPolicy>>;
+
```

Comment:
No type traits?

Reply:
I'm just realizing now that we should probably eliminate the type traits for `managed` as well, or else introduce a memory type constexpr member to the accessor. There's no way to distinguish between managed and pinned memory on the basis of host/device access, which brings us back to the `universal` vocabulary we had been discussing.

Hmmm... I'm very conflicted on this one. On the one hand, the memory type is not really relevant to how the data are accessed, so it feels like it doesn't belong on the accessor type. On the other hand, practically speaking, we may actually really care about the distinction between managed and pinned mdspans in certain cases. Thoughts?

I'm slightly leaning toward just eliminating the managed type traits. That would encourage people to dispatch based on memory accessibility, not the specific memory type. If we later added other memory types, that might be the more sustainable way to go.

Reply:
I think we are facing a conundrum here that I'm unable to come up with a good answer for as of now. While I agree that we need to let users pick from these types when the usability so desires that the difference between managed/pinned matters, what about the cases that don't?
Problems that will work as-is:
1. Algorithms (like NN Descent) which have sections that **only** work with pinned/managed memory. This is easier to solve as these APIs can overload on type-traits of `is_host_accessible && is_device_accessible`
2. `host/device` functions that can work with any kind of memory, so you should be able to pass any variant of `mdspan` (again type-trait of point 1 suffices)

Problems that won't work after PR merge:
1. Current APIs that explicitly use `device_matrix_view`: we should be able to pass `managed/pinned_matrix_view` seamlessly to this but it won't work directly constructing a `managed/pinned_mdarray`, the user will have to use `mdbuffer` and get the right `mdspan` variant

This problem is specifically why I first thought of a `universal_mdspan` but that is an incomplete solution as well. What we need to figure out is if there is a way to implicitly convert `pinned/managed_mdspan` to `device/host_mdspan`. If we can't do it implicitly, then explicitly is easy enough. We can add overloads to `make_device/host_*` that accept any `mdspan` that satisfies the correct type-trait and returns a pure device or host-only view.

The advantage of the explicit function is that we formalize the back-bone of our APIs as device/host/both variants, and give users the tools to interchange our types when possible for APIs that DO NOT provide overloads for all 3. For APIs that already provide device/host overloads, the task then becomes as simple as adding a third overload.

Reply:
Going to mark this as resolved for now, since I think this discussion is really leading us back to the memory type dispatcher, and we can discuss that in its implementation. Happy to revisit anything here after we've brought the dispatcher into the conversation.